[analyzer] Add optin.taint.TaintedDiv checker

dkrupp opened the pull request:
Tainted division operation is separated out from the core.DivideZero checker into the optional optin.taint.TaintedDiv checker. The checker warns when the denominator in a division operation is an attacker controlled value.

dkrupp committed:
Adding optin.taint.TaintedDiv checker

Tainted division operation is separated out from the core.DivideZero
checker into the optional optin.taint.TaintedDiv checker.
The checker warns when the denominator in a division operation
is an attacker controlled value.

dkrupp committed:
formatting fix

NagyDonat commented on clang/lib/StaticAnalyzer/Checkers/DivZeroChecker.cpp:
```diff
@@ -25,16 +25,20 @@ using namespace ento;
 
 public:
+  /// This checker class implements multiple user facing checker
+  enum CheckKind { CK_DivZeroChecker, CK_TaintedDivChecker, CK_NumCheckKinds };
```
Please use `DivideZero` instead of `DivZero` in the name of the `CheckKind` to be consistent with the name of the `core.DivideZero` checker. (The name of source file and the checker class may stay `DivZero`, because they are no longer in 1:1 correspondence with the checker. It's unfortunate that we have these inconsistencies, they are really annoying during the development.)

Also I'd omit the `Checker` suffix from these enum constants -- it doesn't provide any added information.

NagyDonat reviewed (commented):
LGTM overall, I added some minor inline remarks.

Also consider adding a few simple testcases to distinguish the effects of DivideZero and TaintedDiv. It would also be interesting to highlight what happens in situations like

```c
int test(void) {
  int x = getchar(); // or any other taint source
  if (!x)
    return 5 / x;
  return 8;
}
```
(I presume that in this case core.DivideZero will create a bug report, but the new TaintedDiv checker won't.)

steakhal commented:
I had a look, and I don't think I have more to add to the existing review.

steakhal reviewed (commented):
Ah it didn't publish my inline comments. Here they are.

dkrupp commented:
> LGTM overall, I added some minor inline remarks.
> 
> Also consider adding a few simple testcases to distinguish the effects of DivideZero and TaintedDiv. It would also be interesting to highlight what happens in situations like
> 
> ```c
> int test(void) {
>   int x = getchar(); // or any other taint source
>   if (!x)
>     return 5 / x;
>   return 8;
> }
> ```
> 
> (I presume that in this case core.DivideZero will create a bug report, but the new TaintedDiv checker won't.)

Test added.

dkrupp committed:
Fixing review comments

-Making tainted division the report non-fatal so the analysis can continue
-Adding some test cases
-Fixing minor documentation issues

dkrupp committed:
formatting fixes

NagyDonat commented on clang/docs/analyzer/checkers.rst:
```diff
@@ -1288,6 +1288,34 @@ by explicitly marking the ``size`` parameter as sanitized. See the
+This checker warns when the denominator in a division
+operation is a tainted (potentially attacker controlled) value.
+If the attacker can set the denominator to 0, a runtime error can
+be triggered. The checker warns if the analyzer cannot prove
+that the denominator is not 0 and it is a tainted value.
```
```suggestion
be triggered. The checker warns when the denominator is a tainted value and the
analyzer cannot prove that it is not 0.
```
Swap these two conditions to clarify that  "is a tainted value" is not within the "the analyzer cannot prove that" block.

NagyDonat commented on clang/docs/analyzer/checkers.rst:
```diff
@@ -1288,6 +1288,34 @@ by explicitly marking the ``size`` parameter as sanitized. See the
+
+optin.taint.TaintedDiv (C, C++, ObjC)
+"""""""""""""""""""""""""""""""""""""
+This checker warns when the denominator in a division
+operation is a tainted (potentially attacker controlled) value.
+If the attacker can set the denominator to 0, a runtime error can
+be triggered. The checker warns if the analyzer cannot prove
+that the denominator is not 0 and it is a tainted value.
+This warning is more pessimistic than the :ref:`core-DivideZero` checker
+which warns only when it can prove that the denominator is 0.
```
Re-wrap this text to a consistent line length of 80 characters.

NagyDonat commented on clang/docs/analyzer/checkers.rst:
```diff
@@ -1288,6 +1288,34 @@ by explicitly marking the ``size`` parameter as sanitized. See the
+  int vulnerable(int n) {
+    size_t size = 0;
+    scanf("%zu", &size);
+    return n/size; // warn: Division by a tainted value, possibly zero
```
```suggestion
    return n / size; // warn: Division by a tainted value, possibly zero
```
AFAIK coding style guides almost always suggest spaces around binary operators.

NagyDonat commented on clang/docs/analyzer/checkers.rst:
```diff
@@ -1288,6 +1288,34 @@ by explicitly marking the ``size`` parameter as sanitized. See the
+    scanf("%zu", &size);
+    if (!size)
+      return 0;
+    return n/size; // no warning
```
```suggestion
    return n / size; // no warning
```
As above.

NagyDonat commented on clang/include/clang/StaticAnalyzer/Checkers/Checkers.td:
```diff
@@ -1703,6 +1703,12 @@ def TaintedAllocChecker: Checker<"TaintedAlloc">,
   Documentation<HasDocumentation>;
 
+def TaintedDivChecker: Checker<"TaintedDiv">,
+  HelpText<"Check for divisions, where the denominator "
+           "might be 0 as it is a tainted (attacker controlled) value.">,
```
```suggestion
  HelpText<"Check for divisions where the denominator is tainted "
           "(attacker controlled) and might be 0.">,
```
Again, swapping the order clearly clarifies that taintedness is not covered by the "might". (Your wording was also understandable, but more complicated.)

NagyDonat commented on clang/lib/StaticAnalyzer/Checkers/DivZeroChecker.cpp:
```diff
@@ -25,16 +25,20 @@ using namespace ento;
                       llvm::ArrayRef<SymbolRef> TaintedSyms) const;
 
 public:
+  /// This checker class implements multiple user facing checker
```
```suggestion
  /// This checker class implements several user facing checkers
```

NagyDonat commented on clang/test/Analysis/taint-generic.c:
```diff
@@ -20,6 +22,7 @@
 // RUN: not %clang_analyze_cc1 -Wno-pointer-to-int-cast \
 // RUN:   -Wno-incompatible-library-redeclaration -verify %s \
 // RUN:   -analyzer-checker=optin.taint.GenericTaint  \
+// RUN:   -analyzer-checker=optin.taint.TaintedDiv \
```
```suggestion
```
I strongly suspect that there is no reason to add TaintedDiv here -- this not a normal test run, but an intentionally wrong invocation that passes an invalid config file name and verifies the error output with FileCheck.

In fact, I also suspect that you could remove the ExprInspection line below this without changing the meaning of this test. (You should perhaps do it for the sake of clarity.)

dkrupp replied:
I removed the superflous options from other invocations too.

NagyDonat commented on clang/test/Analysis/taint-diagnostic-visitor.c:
```diff
@@ -1,4 +1,4 @@
-// RUN: %clang_cc1 -analyze -analyzer-checker=optin.taint,core,alpha.security.ArrayBoundV2,optin.taint.TaintedAlloc -analyzer-output=text -verify %s
+// RUN: %clang_cc1 -analyze -analyzer-checker=optin.taint,core,alpha.security.ArrayBoundV2,optin.taint.TaintedAlloc,optin.taint.TaintedDiv -analyzer-output=text -verify %s
```
This run line is heavily redundant: you first enable the group `optin.taint`, then you also enable two checkers from it individually.

NagyDonat reviewed (commented):
Looks good overall, I added several suggestions but they mostly tweak the documentation and the comments.

Moreover, I don't see any tests where the separation between  `optin.taint.TaintedDiv` and `core.DivideZero` is tested by enabling one of them and disabling the other. You should probably add a new short test file for this where:
 - taint modeling is active;
 - there are two RUN lines (with different `-verify=...` tags): one that enables `optin.taint.TaintedDiv` and one that enables `core.DivideZero`;
 - there are testcases that perform division by:
   - a value that comes from a tainted source and is known to be zero at the point of division (I know that e.g. in an `if (x == 0)` block `getSVal()` will return a `ConcreteInt` as the value of `x` instead of the (potentially tainted) symbol that was used to represent it earlier -- but it would be still nice to see this situation.)
   - a value that comes from a tainted source and may be either zero or nonzero;
   - a value that comes from a tainted source and is known to be nonzero.

dkrupp committed:
Fixing review comments

-separating test cases for the tainteddiv and the dividezero checkers
-simplifying taint test execution lit test invocations

dkrupp commented:
> Looks good overall, I added several suggestions but they mostly tweak the documentation and the comments.
> 
> Moreover, I don't see any tests where the separation between `optin.taint.TaintedDiv` and `core.DivideZero` is tested by enabling one of them and disabling the other. You should probably add a new short test file for this where:
> 
>     * taint modeling is active;
> 
>     * there are two RUN lines (with different `-verify=...` tags): one that enables `optin.taint.TaintedDiv` and one that enables `core.DivideZero`;
> 
>     * there are testcases that perform division by:
>       
>       * a value that comes from a tainted source and is known to be zero at the point of division (I know that e.g. in an `if (x == 0)` block `getSVal()` will return a `ConcreteInt` as the value of `x` instead of the (potentially tainted) symbol that was used to represent it earlier -- but it would be still nice to see this situation.)
>       * a value that comes from a tainted source and may be either zero or nonzero;
>       * a value that comes from a tainted source and is known to be nonzero.

Thanks. I added divzero-tainted-div-difference.c test file for this.

NagyDonat commented on clang/docs/analyzer/checkers.rst:
```diff
@@ -1288,6 +1288,34 @@ by explicitly marking the ``size`` parameter as sanitized. See the
+This checker warns when the denominator in a division
+operation is a tainted (potentially attacker controlled) value.
+If the attacker can set the denominator to 0, a runtime error can
+be triggered. The checker warns when the denominator is a  tainted
+value  and the analyzer cannot prove that it is not 0. This warning
```
```suggestion
be triggered. The checker warns when the denominator is a tainted
value and the analyzer cannot prove that it is not 0. This warning
```
There were two spaces instead of one at these two locations.

NagyDonat commented on clang/test/Analysis/taint-generic.c:
```diff
@@ -11,16 +12,16 @@
 
-// RUN: not %clang_analyze_cc1 -Wno-pointer-to-int-cast \
-// RUN:   -Wno-incompatible-library-redeclaration -verify %s \
+// RUN: not %clang_analyze_cc1 \
+// RUN:   -verify %s \
```
```suggestion
// RUN: not %clang_analyze_cc1 -verify %s \
```

NagyDonat commented on clang/test/Analysis/taint-generic.c:
```diff
@@ -30,10 +31,9 @@
 // CHECK-INVALID-FILE-SAME:        'justguessit'
 
-// RUN: not %clang_analyze_cc1 -Wno-incompatible-library-redeclaration \
+// RUN: not %clang_analyze_cc1 \
 // RUN:   -verify %s \
```
```suggestion
// RUN: not %clang_analyze_cc1 -verify %s \
```

NagyDonat commented on clang/test/Analysis/taint-generic.c:
```diff
@@ -42,10 +42,9 @@
 // CHECK-ILL-FORMED-SAME:        that expects a valid yaml file: [[MSG]]
 
-// RUN: not %clang_analyze_cc1 -Wno-incompatible-library-redeclaration \
+// RUN: not %clang_analyze_cc1 \
 // RUN:   -verify %s \
```
```suggestion
// RUN: not %clang_analyze_cc1 -verify %s \
```

NagyDonat reviewed (approved):
Looks good to me – _after some very minor cleanup_ ;)

NagyDonat committed:
Minor cleanup of documentation and RUN command line breaks
